engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/unusedimports/UsedSymbolsCollector.java

```diff
@@ -0,0 +1,309 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(UsedSymbolsCollector.class);
+
+  private UsedSymbolsCollector(Module moduleIr, BindingsMap bindingsMap) {
+    this.moduleIr = moduleIr;
+    this.bindingsMap = bindingsMap;
+    this.usedSymbolsBldr = new Builder();
+  }
+
+  static UsedSymbols collect(Module moduleIr, BindingsMap bindingsMap) {
+    var collector = new UsedSymbolsCollector(moduleIr, bindingsMap);
+    return collector.collect();
+  }
+
+  private UsedSymbols collect() {
+    gatherUsedSymbolsFromExports(moduleIr);
+    collectFromRoot(moduleIr);
+    var usedSymbols = usedSymbolsBldr.build();
+    return usedSymbols;
+  }
+
+  /**
+   * Traverses the whole subtree of the given {@code ir} and collects all the {@link Resolution}
+   * metadata. Note that this is an expensive method.
+   */
+  private void collectFromRoot(IR root) {
+    var irsToProcess = new ArrayDeque<IR>();
+    irsToProcess.add(root);
+    while (!irsToProcess.isEmpty()) {
+      var ir = irsToProcess.removeFirst();
+      // Application.Prefix (method calls) are handled specifically. GlobalNames pass assigns
+      // resolution to the first synthetic self argument.
+      if (ir instanceof Application.Prefix app
+          && app.function() instanceof Name.Literal funcLiteral) {
+        if (!app.arguments().isEmpty()) {
+          var selfArg = app.arguments().head();
+          var selfArgResolution =
+              MetadataInteropHelpers.getMetadataOrNull(
+                  selfArg.value(), GlobalNames$.MODULE$, BindingsMap.Resolution.class);
+          if (selfArgResolution != null) {
+            var targetModName = selfArgResolution.target().module().getName();
+            QualifiedName targetSymbolName;
+            switch (selfArgResolution.target()) {
+              case BindingsMap.ResolvedType tp -> {
+                targetSymbolName = targetModName.createChild(tp.tp().name());
+              }
+              case BindingsMap.ResolvedMethod method -> {
+                targetSymbolName = targetModName.createChild(method.methodName());
+              }
+              case BindingsMap.ResolvedModule module -> {
+                // This is most likely call of a module method
+                targetSymbolName = targetModName.createChild(funcLiteral.name());
+              }
+              default -> {
+                targetSymbolName = selfArgResolution.target().qualifiedName();
+              }
+            }
+            addUsedSymbol(targetModName, targetSymbolName);
+            // Add all the children except for the first argument
+            asJava(app.arguments()).stream().skip(1).forEach(irsToProcess::addLast);
+            irsToProcess.addLast(app.function());
+          } else {
+            irsToProcess.addAll(asJava(ir.children()));
+          }
+        } else {
+          irsToProcess.addAll(asJava(ir.children()));
+        }
+      } else {
+        addUsedSymbolForResolution(getGlobalNamesMeta(ir));
```

Comment:
Unfortunate.
- every code that gathers type information is complicated
- when the information is spread around various passes
- I've seen traces of that in docs generator

https://github.com/enso-org/enso/blob/59eb1f2aeb6369fe2abde95f66385a3927064a9e/engine/runtime-compiler/src/main/java/org/enso/compiler/docs/DocsUtils.java#L21

- is there an IR that contains multiple information like this?
- or each IR contains only single definition of these kinds?
- e.g. can we just make sure `BindingsMap.Resolution.class` is available on the `IR`
- and it is enough just to `ir.preorder().filter(_.isInstance[BindingsMap.Resolution])` to get all the info?

Just thinking loud...

Reply:
> is there an IR that contains multiple information like this?

I haven't seen an IR node contain multiple `BindingsMap.Resolution` metadata. Every time, they are associated with a different pass.

> and it is enough just to `ir.preorder().filter(_.isInstance[BindingsMap.Resolution])` to get all the info?

Unfortunately not, since some `BindingsMap.Resolution` metadata is "hidden" inside another metadata, like `TypeSignatures`. In this case, you need to first obtain the `TypeSignatures` metadata, and then do the `preorder` on the `Expression` contained within the `TypeSignatures`, as in https://github.com/enso-org/enso/blob/7d46f06d4bc7e6c952f247868160aef23d72a95f/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/unusedimports/UsedSymbolsCollector.java#L118-L122